File: Que Hacer MDQ v2/Que Hacer MDQ v2/Classes/Tag.m
//
//  Tag.m
//  Que Hacer MDQ v2
//
//  Created by Lucas on 3/17/15.
//  Copyright (c) 2015 Globant iOS MDQ. All rights reserved.
//

#import "Tag.h"
#import "Activity.h"


@implementation Tag

@dynamic desc;
@dynamic tagId;
@dynamic name;
@dynamic activity;

+(Tag*)instanceFromDictionary:(NSDictionary*)aDictionary
{
    //TO BE COMPLETED
    return [[Tag alloc] init];
}

+(Tag*)persistentInstanceFromDictionary:(NSDictionary*)aDictionary
{
    /*
     {
     "id": "1",
     "name": "General",
     "description": "Actividades sin clasificacion"
     },
     */
    NSManagedObjectContext* context = [[CoreDataHelper sharedInstance] managedObjectContext];
    Tag* tag = [NSEntityDescription insertNewObjectForEntityForName:@"Tag" inManagedObjectContext:context];
    tag.tagId = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"id"]];
    
    if ([[aDictionary objectForKey:@"name"] isKindOfClass:[NSNull class]])
    {
        tag.name = @"";
    }
    else
    {
        tag.name = [aDictionary objectForKey:@"name"];
    }
    
    if ([[aDictionary objectForKey:@"description"] isKindOfClass:[NSNull class]])
    {
        tag.desc = @"";
    }
    else
    {
        tag.desc = [aDictionary objectForKey:@"description"];
    }
    
    NSError* error;
    [context save:&error];
    if (error)
    {
        NSLog(@"Something went wrong when saving the tag with id %@",tag.tagId);
        NSLog(@"The error description is : %@",error.localizedDescription);
        return nil;
    }
    return tag;
    
}


@end

File: Que Hacer? MdQ/Que Hacer? MdQ/Que Hacer? MdQ/Classes/DataManager/Tag.swift
//
//  Tag.swift
//  Que Hacer? MdQ
//
//  Created by Federico Gonzalez on 2/11/15.
//  Copyright (c) 2015 Globant iOS MDQ. All rights reserved.
//

import Foundation
import CoreData

@objc(Tag)
class Tag: NSManagedObject {

    @NSManaged var id: Int
    @NSManaged var name: String
    @NSManaged var desc: String
    
    @NSManaged var activities: NSSet
    
    //To get a Tag object by its id
    //Returns nil with console error when can't find it or if there is more than 1 result
    class func getTag(tagId: Int) -> Tag? {
        
        var context: NSManagedObjectContext = DataAccessHelper.sharedInstance().managedObjectContext
        let fetchRequest = NSFetchRequest(entityName: "Tag")
        
        fetchRequest.returnsObjectsAsFaults = false
        fetchRequest.includesPropertyValues = false  
        fetchRequest.predicate = NSPredicate(format: "id == %d", tagId)
        
        var error: NSError? = nil
        var matches: NSArray = context.executeFetchRequest(fetchRequest, error: &error)!
        
        if (matches.count == 1) {
            return matches[0] as? Tag
        } else if (matches.count > 1){
            NSLog("Error while retrieving unique area, more than one result.")
            NSLog("%@", matches)
        } else {
            NSLog("Couldn't find tag with id = %d", tagId)
        }
        
        return nil
    }
    
}